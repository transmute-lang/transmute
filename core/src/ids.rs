macro_rules! make_id {
    ($name:ident) => {
        #[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Default)]
        pub struct $name {
            id: usize,
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.id)
            }
        }

        impl std::fmt::Debug for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}({})", stringify!($name), self.id)
            }
        }

        impl From<usize> for $name {
            fn from(id: usize) -> Self {
                Self { id }
            }
        }

        impl From<$name> for usize {
            fn from(id: $name) -> Self {
                id.id
            }
        }

        impl std::ops::Add<usize> for $name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                Self::from(self.id + rhs)
            }
        }

        impl std::ops::Add<usize> for &$name {
            type Output = $name;

            fn add(self, rhs: usize) -> Self::Output {
                $name::from(self.id + rhs)
            }
        }
    };
}

pub(crate) use make_id;

make_id!(IdentId);
make_id!(IdentRefId);
make_id!(ExprId);
make_id!(StmtId);
make_id!(SymbolId);
make_id!(TypeId);
make_id!(FunctionId);
make_id!(StructId);
make_id!(TypeDefId);

#[macro_export]
macro_rules! id {
    ($e:expr) => {
        <_ as Into<usize>>::into($e)
    };
}

pub use id;
