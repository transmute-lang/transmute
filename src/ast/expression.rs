use crate::ast::literal::Literal;
use crate::ast::operators::{BinaryOperator, UnaryOperator};
use crate::ast::Visitor;
use crate::lexer::{Location, Span};

#[derive(Debug, PartialEq)]
pub struct Expression {
    kind: ExpressionKind,
}

impl Expression {
    pub fn new(kind: ExpressionKind) -> Self {
        Self { kind }
    }

    pub fn kind(&self) -> &ExpressionKind {
        &self.kind
    }

    pub fn location(&self) -> &Location {
        match &self.kind {
            ExpressionKind::Literal(l) => l.location(),
            ExpressionKind::Binary(e, _, _) => e.location(),
            ExpressionKind::Unary(o, _) => o.location(),
        }
    }

    pub fn span(&self) -> Span {
        match &self.kind {
            ExpressionKind::Literal(l) => l.span().clone(),
            ExpressionKind::Binary(l, _, r) => l.span().extend_to(&r.span()),
            ExpressionKind::Unary(o, e) => o.span().extend_to(&e.span()),
        }
    }

    pub fn accept<V, R>(&self, visitor: &mut V) -> R
    where
        V: Visitor<R>,
    {
        visitor.visit_expression(self)
    }
}

impl From<Literal> for Expression {
    fn from(literal: Literal) -> Self {
        Self {
            kind: ExpressionKind::Literal(literal),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum ExpressionKind {
    Literal(Literal),
    Binary(Box<Expression>, BinaryOperator, Box<Expression>),
    Unary(UnaryOperator, Box<Expression>),
}
