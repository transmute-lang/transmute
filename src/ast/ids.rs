macro_rules! make_id {
    ($name:ident) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
        pub struct $name {
            id: usize,
        }

        impl $name {
            pub fn id(&self) -> usize {
                self.id
            }
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.id)
            }
        }

        impl From<usize> for $name {
            fn from(id: usize) -> Self {
                Self { id }
            }
        }
    };
}

pub(crate) use make_id;

make_id!(IdentId);
make_id!(IdentRefId);
make_id!(ExprId);
make_id!(StmtId);
make_id!(SymbolId);
