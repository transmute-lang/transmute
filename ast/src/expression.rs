use crate::literal::Literal;
use crate::operators::{BinaryOperator, UnaryOperator};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
}

impl Expression {
    pub fn new(id: ExprId, kind: ExpressionKind, span: Span) -> Self {
        Self { id, kind, span }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Binary(ExprId, BinaryOperator, ExprId),
    Unary(UnaryOperator, ExprId),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    /// A dummy expression kind, inserted by the parser when the expression could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}
