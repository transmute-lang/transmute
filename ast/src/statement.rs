use crate::annotation::Annotation;
use crate::identifier::Identifier;
use transmute_core::ids::{ExprId, IdentRefId, InputId, StmtId, TypeDefId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Statement {
    pub id: StmtId,
    pub kind: StatementKind,
    pub span: Span,
}

impl Statement {
    pub fn new(id: StmtId, kind: StatementKind, span: Span) -> Self {
        Self { id, kind, span }
    }

    pub fn as_namespace(&self) -> (&Identifier, InputId, &[StmtId]) {
        match &self.kind {
            StatementKind::Namespace(identifier, input_id, stmts) => (identifier, *input_id, stmts),
            _ => panic!("{:?} is not a namespace", self),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum StatementKind {
    Expression(ExprId),
    Let(Identifier, ExprId),
    Ret(Option<ExprId>, RetMode),
    LetFn(Identifier, Vec<Annotation>, Vec<Parameter>, Return, ExprId),
    Struct(Identifier, Vec<Annotation>, Vec<Field>),
    Annotation(Identifier),
    Use(Vec<IdentRefId>),
    Namespace(
        /// Namespace identifier
        Identifier,
        /// `InputId` this namespace is coming from
        InputId,
        /// Statements included in this namespace
        Vec<StmtId>,
    ),
}

/// AST only produces `Ret` with `RetMode::Explicit` but having that information in the AST nodes
/// helps in the hir transformations (see `ImplicitRetConverter::visit_expression`).
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RetMode {
    Explicit,
    Implicit,
}

impl RetMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            RetMode::Explicit => "explicit",
            RetMode::Implicit => "implicit",
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct TypeDef {
    pub kind: TypeDefKind,
    pub span: Span,
}

impl TypeDef {
    pub fn identifier_ref_id(&self) -> &[IdentRefId] {
        match &self.kind {
            TypeDefKind::Simple(ident_ref_id) => ident_ref_id,
            TypeDefKind::Array(_, _) => todo!(),
            _ => panic!("no identifier_ref_id for TypeDefKind::Dummy"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypeDefKind {
    Simple(Vec<IdentRefId>),
    Array(TypeDefId, usize),
    /// A dummy type kind, inserted by the parser when the type could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Parameter {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Parameter {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Return {
    pub type_def_id: Option<TypeDefId>,
}

impl Return {
    pub fn none() -> Self {
        Self { type_def_id: None }
    }

    pub fn some(type_def_id: TypeDefId) -> Self {
        Self {
            type_def_id: Some(type_def_id),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    pub identifier: Identifier,
    pub type_def_id: TypeDefId,
    pub span: Span,
}

impl Field {
    pub fn new(identifier: Identifier, type_def_id: TypeDefId, span: Span) -> Self {
        Self {
            identifier,
            type_def_id,
            span,
        }
    }
}
