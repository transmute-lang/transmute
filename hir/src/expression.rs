use crate::literal::{Literal, LiteralKind};
use crate::typed::{Typed, TypedState, Untyped};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId, TypeId};
use transmute_core::span::Span;
use transmute_nst::nodes::Expression as NstExpression;
use transmute_nst::nodes::ExpressionKind as NstExpressionKind;
use transmute_nst::nodes::Target as NstTarget;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression<T>
where
    T: TypedState,
{
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
    typed: T,
}

impl<T> Expression<T>
where
    T: TypedState,
{
    pub fn collect_ident_ref_ids(&self) -> Vec<IdentRefId> {
        match &self.kind {
            ExpressionKind::Assignment(Target::Direct(ident_ref_id), _) => {
                vec![*ident_ref_id]
            }
            ExpressionKind::Literal(lit) => match &lit.kind {
                LiteralKind::Identifier(ident_ref_id) => vec![*ident_ref_id],
                _ => Default::default(),
            },
            ExpressionKind::Access(_, ident_ref_id) => vec![*ident_ref_id],
            ExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                let mut ident_ref_ids = fields
                    .iter()
                    .map(|(ident_ref_id, _)| *ident_ref_id)
                    .collect::<Vec<_>>();
                ident_ref_ids.push(*ident_ref_id);
                ident_ref_ids
            }
            _ => Default::default(),
        }
    }
}

impl Expression<Untyped> {
    pub fn typed(self, ty: TypeId) -> Expression<Typed> {
        Expression {
            id: self.id,
            kind: self.kind,
            span: self.span,
            typed: Typed(ty),
        }
    }
}

impl From<NstExpression> for Expression<Untyped> {
    fn from(value: NstExpression) -> Self {
        Self {
            id: value.id,
            span: value.span.clone(),
            kind: match value.kind {
                NstExpressionKind::Assignment(target, expr_id) => {
                    ExpressionKind::Assignment(Target::from(target), expr_id)
                }
                NstExpressionKind::If(cond, true_branch, false_branch) => {
                    ExpressionKind::If(cond, true_branch, false_branch)
                }
                NstExpressionKind::Literal(literal) => {
                    ExpressionKind::Literal(Literal::from(literal))
                }
                NstExpressionKind::Access(expr_id, ident_ref_id) => {
                    ExpressionKind::Access(expr_id, ident_ref_id)
                }
                NstExpressionKind::FunctionCall(expr_id, expr_ids) => {
                    ExpressionKind::FunctionCall(expr_id, expr_ids)
                }
                NstExpressionKind::While(cond, expr_id) => ExpressionKind::While(cond, expr_id),
                NstExpressionKind::Block(expr_ids) => ExpressionKind::Block(expr_ids),
                NstExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                    ExpressionKind::StructInstantiation(ident_ref_id, fields)
                }
                NstExpressionKind::ArrayInstantiation(values) => {
                    ExpressionKind::ArrayInstantiation(values)
                }
                NstExpressionKind::ArrayAccess(expr, index) => {
                    ExpressionKind::ArrayAccess(expr, index)
                }
            },
            typed: Untyped,
        }
    }
}

impl Expression<Typed> {
    pub fn resolved_type_id(&self) -> TypeId {
        self.typed.0
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Access(ExprId, IdentRefId),
    FunctionCall(ExprId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    ArrayInstantiation(Vec<ExprId>),
    ArrayAccess(ExprId, ExprId),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}

impl From<NstTarget> for Target {
    fn from(value: NstTarget) -> Self {
        match value {
            NstTarget::Direct(ident_ref_id) => Self::Direct(ident_ref_id),
            NstTarget::Indirect(expr_id) => Self::Indirect(expr_id),
        }
    }
}
