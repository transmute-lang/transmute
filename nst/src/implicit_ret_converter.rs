use transmute_ast::expression::{Expression, ExpressionKind, Target};
use transmute_ast::statement::{RetMode, Statement, StatementKind};
use transmute_core::ids::{ExprId, StmtId};
use transmute_core::vec_map::VecMap;

#[derive(Default)]
pub struct ImplicitRetResolver {
    statements: VecMap<StmtId, Statement>,
    expressions: VecMap<ExprId, Expression>,
}

pub struct Output {
    pub statements: VecMap<StmtId, Statement>,
    pub expressions: VecMap<ExprId, Expression>,
}

impl ImplicitRetResolver {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn resolve(
        mut self,
        root_statement: StmtId,
        mut statements: VecMap<StmtId, Statement>,
        mut expressions: VecMap<ExprId, Expression>,
    ) -> Output {
        self.statements.resize(statements.len());
        self.expressions.resize(expressions.len());

        #[cfg(debug_assertions)]
        let (initial_statements_len, initial_expressions_len) =
            (statements.len(), expressions.len());

        let statement = statements.remove(root_statement).unwrap();
        self.visit_statement(
            &mut statements,
            &mut expressions,
            statement,
            0,
            false,
            false,
        );

        // we may have skipped the visit of statements or expressions due to the fact that their
        // parent was unreachable, and thus its children were not visited... we transfer them now.
        statements.into_iter().for_each(|(stmt_id, stmt)| {
            self.statements.insert(stmt_id, stmt);
        });
        expressions.into_iter().for_each(|(expr_id, expr)| {
            self.expressions.insert(expr_id, expr);
        });

        #[cfg(debug_assertions)]
        {
            debug_assert!(self.statements.len() >= initial_statements_len);
            debug_assert!(!self.statements.has_holes());
            debug_assert!(self.expressions.len() >= initial_expressions_len);
            debug_assert!(!self.expressions.has_holes());
        }

        Output {
            statements: self.statements,
            expressions: self.expressions,
        }
    }

    /// returns true if all nested paths explicitly return
    /// the depth is used to know if the expression is the function's last expression
    fn visit_expression(
        &mut self,
        statements: &mut VecMap<StmtId, Statement>,
        expressions: &mut VecMap<ExprId, Expression>,
        expression: Expression,
        depth: usize,
        unreachable: bool,
    ) -> bool {
        if unreachable {
            // we don't visit unreachable expression, possibly missing implicit exit points. as the
            // expressions are not reachable anyway, we ar safe
            self.expressions.insert(expression.id, expression);
            return false;
        }

        match &expression.kind {
            ExpressionKind::Assignment(target, expr) => {
                let target_expression = match target {
                    Target::Direct(_) => None,
                    Target::Indirect(target) => Some(expressions.remove(*target).unwrap()),
                };

                let value_expression = expressions.remove(*expr).unwrap();

                self.expressions.insert(expression.id, expression);

                let target_expression_returns = target_expression
                    .map(|expression| {
                        self.visit_expression(
                            statements,
                            expressions,
                            expression,
                            depth + 1,
                            unreachable,
                        )
                    })
                    .unwrap_or(false);

                let value_expression_returns = self.visit_expression(
                    statements,
                    expressions,
                    value_expression,
                    depth + 1,
                    unreachable,
                );

                value_expression_returns || target_expression_returns
            }
            ExpressionKind::If(cond, true_branch, false_branch) => {
                let cond_expression = expressions.remove(*cond).unwrap();
                let condition_always_returns = self.visit_expression(
                    statements,
                    expressions,
                    cond_expression,
                    depth + 1,
                    unreachable,
                );

                let true_branch_expression = expressions.remove(*true_branch).unwrap();
                let false_branch_expression =
                    false_branch.map(|expr_id| expressions.remove(expr_id).unwrap());

                let branches_always_return = if condition_always_returns {
                    // not that not visiting the true branch and the false branch expression
                    // prevents us to find potential implicit exit points there, but since  the
                    // condition always returns anyway we're safe
                    self.expressions
                        .insert(true_branch_expression.id, true_branch_expression);

                    if let Some(false_branch_expression) = false_branch_expression {
                        self.expressions
                            .insert(false_branch_expression.id, false_branch_expression);
                    }

                    // we don't know whether both branches always return, and we don't care as the
                    // condition always returns anyway
                    false
                } else {
                    let true_branch_always_return = self.visit_expression(
                        statements,
                        expressions,
                        true_branch_expression,
                        depth + 1,
                        unreachable,
                    );
                    let false_branch_always_return = false_branch_expression
                        .map(|false_branch_expression| {
                            self.visit_expression(
                                statements,
                                expressions,
                                false_branch_expression,
                                depth + 1,
                                unreachable,
                            )
                        })
                        .unwrap_or(false);

                    true_branch_always_return && false_branch_always_return
                };

                self.expressions.insert(expression.id, expression);

                condition_always_returns || branches_always_return
            }
            ExpressionKind::Literal(_) => {
                self.expressions.insert(expression.id, expression);
                false
            }
            ExpressionKind::Binary(..) | ExpressionKind::Unary(..) => {
                panic!("operators must be converted to functions");
            }
            ExpressionKind::Access(expr_id, _) => {
                let target_expression = expressions.remove(*expr_id).unwrap();

                self.expressions.insert(expression.id, expression);

                self.visit_expression(
                    statements,
                    expressions,
                    target_expression,
                    depth + 1,
                    unreachable,
                )
            }
            ExpressionKind::FunctionCall(_, params) => {
                let mut some_param_always_returns = false;

                for param in params {
                    let param_expression = expressions.remove(*param).unwrap();
                    some_param_always_returns = some_param_always_returns
                        || self.visit_expression(
                            statements,
                            expressions,
                            param_expression,
                            depth + 1,
                            unreachable || some_param_always_returns,
                        );
                }

                self.expressions.insert(expression.id, expression);

                some_param_always_returns
            }
            ExpressionKind::While(cond, expr) => {
                let cond_expression = expressions.remove(*cond).unwrap();
                let condition_always_returns = self.visit_expression(
                    statements,
                    expressions,
                    cond_expression,
                    depth + 1,
                    unreachable,
                );

                let block_expression = expressions.remove(*expr).unwrap();

                let block_always_returns = if condition_always_returns {
                    // not that not visiting the block expression prevents us to find potential
                    // implicit exit points there, but since the condition always returns
                    // anyway we're safe
                    self.expressions
                        .insert(block_expression.id, block_expression);

                    // we don't know whether the body always return, and we don't care as the
                    // condition always returns anyway
                    false
                } else {
                    self.visit_expression(
                        statements,
                        expressions,
                        block_expression,
                        depth + 1,
                        unreachable || condition_always_returns,
                    )
                };

                self.expressions.insert(expression.id, expression);

                condition_always_returns || block_always_returns
            }
            ExpressionKind::Block(stmts) if stmts.is_empty() => {
                if depth > 0 {
                    // we're not the root block of a function, we don't want to insert an implicit
                    // return. moreover, as the block is empty, it cannot always return. hence, we
                    // just insert the expression, and return `false`...

                    self.expressions.insert(expression.id, expression);

                    return false;
                }

                let stmt_id = self.statements.next_index();

                self.statements.push(Statement::new(
                    stmt_id,
                    StatementKind::Ret(None, RetMode::Implicit),
                    expression.span.clone(),
                ));

                self.expressions.insert(
                    expression.id,
                    Expression {
                        id: expression.id,
                        kind: ExpressionKind::Block(vec![stmt_id]),
                        span: expression.span,
                    },
                );

                true
            }
            ExpressionKind::Block(stmts) => {
                let stmts_len = stmts.len();
                let mut ret = false;
                for (i, stmt) in stmts.iter().enumerate() {
                    let last_statement = depth == 0 && i == stmts_len - 1 && !unreachable;

                    let statement = statements.remove(*stmt).unwrap();
                    if self.visit_statement(
                        statements,
                        expressions,
                        statement,
                        depth,
                        unreachable || ret,
                        last_statement,
                    ) {
                        ret = true;
                    }
                }

                if depth == 0 && !ret {
                    let stmt_id = self.statements.next_index();
                    self.statements.insert(
                        stmt_id,
                        Statement {
                            id: stmt_id,
                            kind: StatementKind::Ret(None, RetMode::Implicit),
                            // todo:ux can we do better?
                            span: Default::default(),
                        },
                    );

                    let mut stmts = stmts.clone();
                    stmts.push(stmt_id);

                    self.expressions.insert(
                        expression.id,
                        Expression {
                            id: expression.id,
                            kind: ExpressionKind::Block(stmts),
                            span: expression.span,
                        },
                    );
                } else {
                    self.expressions.insert(expression.id, expression);
                }

                ret
            }
            ExpressionKind::StructInstantiation(_, fields) => {
                let mut always_return = false;
                for (_, expr_id) in fields {
                    let field_expression = expressions.remove(*expr_id).unwrap();

                    let expression_returns = self.visit_expression(
                        statements,
                        expressions,
                        field_expression,
                        depth + 1,
                        unreachable || always_return,
                    );

                    always_return = always_return || expression_returns;
                }

                self.expressions.insert(expression.id, expression);

                always_return
            }
            ExpressionKind::ArrayInstantiation(values) => {
                let mut always_return = false;
                for expr_id in values {
                    let value_expression = expressions.remove(*expr_id).unwrap();

                    let expression_returns = self.visit_expression(
                        statements,
                        expressions,
                        value_expression,
                        depth + 1,
                        unreachable || always_return,
                    );

                    always_return = always_return || expression_returns;
                }

                self.expressions.insert(expression.id, expression);

                always_return
            }
            ExpressionKind::ArrayAccess(base_expr_id, index_expr_id) => {
                let expr = expressions.remove(*base_expr_id).unwrap();
                let mut always_return =
                    self.visit_expression(statements, expressions, expr, depth + 1, false);

                let index = expressions.remove(*index_expr_id).unwrap();
                always_return = always_return
                    || self.visit_expression(
                        statements,
                        expressions,
                        index,
                        depth + 1,
                        always_return,
                    );

                self.expressions.insert(expression.id, expression);

                always_return
            }
            ExpressionKind::Dummy => {
                panic!("should not compute exit points of an invalid source code")
            }
        }
    }

    /// returns true if all nested paths explicitly return
    fn visit_statement(
        &mut self,
        statements: &mut VecMap<StmtId, Statement>,
        expressions: &mut VecMap<ExprId, Expression>,
        statement: Statement,
        depth: usize,
        unreachable: bool,
        last: bool,
    ) -> bool {
        if unreachable {
            // we don't visit unreachable expression, possibly missing implicit exit points. as the
            // expressions are not reachable anyway, we are safe
            self.statements.insert(statement.id, statement);
            return false;
        }

        match &statement.kind {
            StatementKind::Expression(expr_id) => {
                let expression = expressions.remove(*expr_id).unwrap();
                let always_returns = self.visit_expression(
                    statements,
                    expressions,
                    expression,
                    depth + 1,
                    unreachable,
                );

                if last && !always_returns {
                    self.statements.insert(
                        statement.id,
                        Statement {
                            id: statement.id,
                            kind: StatementKind::Ret(Some(*expr_id), RetMode::Implicit),
                            span: statement.span,
                        },
                    );
                    true
                } else {
                    self.statements.insert(statement.id, statement);
                    always_returns
                }
            }
            StatementKind::Let(_, expr) => {
                let expression = expressions.remove(*expr).unwrap();

                self.statements.insert(statement.id, statement);

                self.visit_expression(statements, expressions, expression, depth + 1, unreachable)
            }
            StatementKind::Ret(Some(expr), _) => {
                let expr = expressions.remove(*expr).unwrap();
                self.visit_expression(statements, expressions, expr, depth + 1, unreachable);

                self.statements.insert(statement.id, statement);

                true
            }
            StatementKind::Ret(None, _) => {
                self.statements.insert(statement.id, statement);
                true
            }
            StatementKind::LetFn(_, _, _, _, expr) => {
                let expr = expressions.remove(*expr).unwrap();
                self.visit_expression(statements, expressions, expr, 0, unreachable);

                self.statements.insert(statement.id, statement);

                false
            }
            StatementKind::Struct(_, _, _)
            | StatementKind::Annotation(_)
            | StatementKind::Use(_) => {
                self.statements.insert(statement.id, statement);
                false
            }
            StatementKind::Namespace(_, _, stmts) => {
                for stmt in stmts {
                    let statement = statements.remove(*stmt).unwrap();
                    self.visit_statement(statements, expressions, statement, 0, false, false);
                }
                self.statements.insert(statement.id, statement);
                false
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use transmute_ast::lexer::Lexer;
    use transmute_ast::parser::Parser;
    use transmute_ast::pretty_print::Options;
    use transmute_ast::CompilationUnit;
    use transmute_core::ids::InputId;

    macro_rules! t {
        ($name:ident, $src:expr) => {
            #[test]
            fn $name() {
                let mut compilation_unit = CompilationUnit::default();

                Parser::new(
                    &mut compilation_unit,
                    None,
                    Lexer::new(InputId::from(0), $src),
                )
                .parse();
                let mut ast = compilation_unit.into_ast().unwrap();

                // move out of ast without taking owership
                let mut statements = VecMap::new();
                for k in ast.statements.keys().into_iter() {
                    statements.insert(k, ast.statements.remove(k).unwrap());
                }

                let mut expressions = VecMap::new();
                for k in ast.expressions.keys().into_iter() {
                    expressions.insert(k, ast.expressions.remove(k).unwrap());
                }

                let explicit_rets =
                    ImplicitRetResolver::new().resolve(ast.root, statements, expressions);

                let expressions = explicit_rets.expressions;
                let statements = explicit_rets.statements;

                // move back into the ast
                for s in statements.into_iter() {
                    ast.statements.insert(s.0, s.1);
                }
                for e in expressions.into_iter() {
                    ast.expressions.insert(e.0, e.1);
                }

                let mut w = String::new();
                let mut options = Options::default();
                options.display_implicit_ret = true;
                ast.pretty_print(&options, &mut w).unwrap();

                assert_snapshot!(w);
            }
        };
    }

    t!(
        explicit_return,
        r#"
            let f(): number = {
                ret 42;
            }
        "#
    );

    t!(
        explicit_void_return,
        r#"
            let f() = {
                ret;
            }
        "#
    );

    t!(
        implicit_void_return,
        r#"
            let f() = {
            }
        "#
    );

    t!(
        implicit_return_simple,
        r#"
            let f(): number = {
                41;
                42;
            }
        "#
    );

    t!(
        implicit_return_if_full,
        r#"
            let f(): number = {
                if c {
                    42;
                }
                else {
                    43;
                }
            }
        "#
    );

    t!(
        implicit_return_if_full_with_explicit,
        r#"
            let f(): number = {
                if c {
                    ret 42;
                }
                else {
                    43;
                }
            }
        "#
    );

    t!(
        implicit_return_if,
        r#"
            let f(): number = {
                if c {
                    42;
                }
                43;
            }
        "#
    );

    t!(
        implicit_return_struct,
        r#"
            struct S { x: number, y: number }
            let f(): number = {
                S {
                    x: if true { ret 1; },
                    y: 1
                };
            }
        "#
    );

    t!(
        nested_function,
        r#"
            let f() {
                let g() {
                    true;
                };
                1;
            }
        "#
    );

    t!(
        while_condition_returns,
        "while if true { ret 42; } else { ret 43; } {}"
    );

    t!(
        if_condition_returns,
        "if if true { ret 42; } else { ret 43; } {}"
    );

    t!(
        struct_access,
        "struct S { field: number } let s = S { field: 1 }; s.field = false;"
    );

    t!(
        implicit_return_after_statement,
        "let f(n: number) { let a = 1; }"
    );

    t!(if_empty_branches, "let f() = { if true { } else { }; }");

    t!(while_empty_branches, "let f() = { while true { }; }");

    t!(
        array_value_always_ret,
        "let f(): number { let a = [1, if true { ret 2; } else { ret 3; }, 4]; 5; }"
    );
    t!(array, "let f(): number { let a = [1, 2, 4]; 5; }");

    t!(array_access_no_ret, "let f() { a[1]; }");
    t!(
        array_access_ret,
        "let f() { a[ if true { ret 1; } else { ret 2; } ]; }"
    );
    t!(namespace, "namespace name { let f(): number { 1; } }");
}
