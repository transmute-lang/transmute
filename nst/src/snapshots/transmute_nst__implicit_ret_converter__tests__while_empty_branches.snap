---
source: nst/src/implicit_ret_converter.rs
expression: w
---
let f() = {
  ret while true {
  };
}
